solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -368,10 +368,152 @@ private StatefulCollectionWatch compute(
+            log.debug("PRS node event : {}", event.getType());
+          }
+
+          PerReplicaStates prs = collectionState.getPerReplicaStates();
+          if (prs != null) {
+            String stateStr = path.substring(collectionPath.length() + 1);
+            PerReplicaStates.State newState = PerReplicaStates.State.parse(stateStr);
+            PerReplicaStates.State oldState = prs.states.get(newState.replica);
+            if (event.getType() == EventType.NodeCreated) {
+              if (oldState != null && newState.version < oldState.version) {
+                // we got a notification out of order? . Shouldn't happen
+                if (log.isTraceEnabled()) {
+                  log.trace("newState {} < oldState {}", newState, oldState);
+                }
+                return;
+              }
+              if (oldState == null) {
+                // the state does not exist now. fetch everything
+
+                if (log.isTraceEnabled()) {
+                  log.trace("fresh replica, force fetch all {}", collectionPath);
+                }
+
+                prs = PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null);
+              } else {
+                // the PRS entry is already available, need to do an update
+                Stat stat = null;
+                try {
+                  // get the stat of this child node
+                  stat = zkClient.exists(path, null, true);
+                } catch (Exception e) {
+                  throw new RuntimeException(e);
+                }
+                if (stat == null) {
+                  // this got deleted so soon. nothing to do
+                  return;
+                }
+
+                if (log.isTraceEnabled()) {
+                  log.trace("PRS insert {}, v:{}", newState, stat.getCzxid());
+                }
+                prs = prs.insert(newState, stat.getCzxid());
+                if (prs == null) {
+                  // something went wrong
+                  prs = PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null);
+                }
+              }
+            } else if (event.getType() == EventType.NodeDeleted) {
+              if (oldState == null) {
+                // This is already removed (unlikely, but)
+                return;
+              }
+              if (newState.version < oldState.version) {
+                if (log.isTraceEnabled()) {
+                  log.trace("PRS in-place remove {}", path);
+                }
+                // removed the duplicate without modifying the PRS Object
+                oldState.removeDuplicate(newState);
+                return;
+              } else {
+                if (log.isTraceEnabled()) {
+                  log.trace("Replica:  {} delete, force fetch", path);
+                }
+                // a replica got removed. we can't get the pzxid reliably. fetch everything
+                prs = PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null);
```

Comment:
similar, can we get stat of `state.json` and use cversion of it?

Reply:
`cversion` is not used anymore at all. it's unreliable